renku/core/models/provenance/provenance_graph.py

```diff
@@ -149,6 +169,39 @@ def query(self, query):
         self._create_rdf_graph()
         return self._graph.query(query)
 
+    def get_subgraph(self, paths):
+        """Return a subgraph that generates specific paths."""
+        starting_activities = []
+        for path in paths:
+            activity = self._get_latest_activity(path)
+            if activity:
+                starting_activities.append(activity)
+
+        if not starting_activities:
+            return self._create_graph()
+
+        todo = deque(starting_activities)
+        activities = []
+
+        while todo:
+            activity = todo.popleft()
+            activities.append(activity)
+            for usage in activity.qualified_usage:
+                parent_activity = self._get_latest_activity(entity_id=usage.entity._id)
+                if not parent_activity:
+                    continue
+                activities.append(parent_activity)
+                todo.append(parent_activity)
+
+        return self._create_graph(activities=activities)
+
+    def _get_latest_activity(self, path=None, entity_id=None):
+        """Return the latest activity that generated a path or entity."""
+        for activity in reversed(self.activities):
+            for generation in activity.generated:
+                if generation.entity.path == path or generation.entity._id == entity_id:
```

Comment:
This would fail if I specify a path of `dir/file1` but the activity had a directory output and generated just `dir/`, no? Since the generation would be a Collection, but with the file in one of its `.entities`.

Even if it works, we should add a test for this case, as things often break with directory outputs.

Reply:
also the opposite case if a user specifies a folder path but the latest activity changed a file in the folder.